### PR TITLE
[MultilevelMonteCarloApplication] Added serialzation, adaptive refinement and MultilevelMonteCarlo class

### PR DESCRIPTION
**Pull request by riccardotosi**
Added adaptive refinement based on the hessian metric.
Cleaning of the application adding the use of a Multilevel Monte Carlo class.
Added the Kratos StreamSerializer object to avoid reading multiple times the parameters and the model part.

TODO: split the ExecuteMultilevelMonteCarloAnalisys_Task task into a smaller task that takes into account a single call of the solutor simulation.Run() and a single adaptive refinement.

### CODE REVIEW

File: kratos/tests/test_kratos_parameters.py [hunk 1]
```diff
@@ -1,666 +0,0 @@
-﻿from __future__ import print_function, absolute_import, division
```
Comment (philbucher) [inline kratos/tests/test_kratos_parameters.py]: this must not be deleted!

**Review by philbucher — CHANGES_REQUESTED** [review]
if yo want to make changes outside of your application the please make separate PRs for those

File: applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py [hunk 2]
```diff
@@ -0,0 +1,781 @@
+from __future__ import absolute_import, division # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
+import numpy as np
+import KratosMultiphysics
+import time
+import copy
+
+# Import exaqute
+from exaqute.ExaquteTaskPyCOMPSs import *   # to exequte with pycompss
```
Comment (RiccardoRossi) [inline applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py]: typo: "execute"

File: applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py [hunk 3]
```diff
@@ -0,0 +1,781 @@
...
+        '''rates_error : dictionary containing the values of the parameters
+        calpha : coefficient of the function maximizing bias
+        alpha  : exponent of the function maximizing bias
+        cbeta  : coefficient of the function maximizing statistical error
+        beta   : exponent of the function maximizing statistical error
+        cgamma : coefficient of the function maximizing cost
+        gamma  : exponent of the function maximizing cost
+        '''
+        self.rates_error = {"calpha":None, "alpha":None, "cbeta":None, "beta":None, "cgamma":None, "gamma":None}
+        '''mesh_parameters : reciprocal of minimal mesh size'''
```
Comment (RiccardoRossi) [inline applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py:262]: for single line comments, you'd best use the "#" symbol

also (for the future, don't need to change here) you can do something like

~~~~py 
    self.mesh_parameters = []   #size_mesh : minimal mesh size

File: applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py [hunk 4]
```diff
@@ -0,0 +1,781 @@
...
+        self.rates_error = {"calpha":None, "alpha":None, "cbeta":None, "beta":None, "cgamma":None, "gamma":None}
+        '''mesh_parameters : reciprocal of minimal mesh size'''        
+        self.mesh_parameters = []
+        '''size_mesh : minimal mesh size'''
+        self.sizes_mesh = []
+        '''BayesianVariance : Bayesian variance'''
+        self.BayesianVariance = []
+        '''number_iterations : theoretical number of iterations the MLMC algorithm will perform'''
+        self.number_iterations_iE = None
+        '''convergence : boolean variable defining if MLMC algorithm is convergenced'''
```
Comment (RiccardoRossi) [inline applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py:270]: typo convergenced

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
blocking since the test_kratos_parameters cannot be deleted!!

File: kratos/tests/test_kratos_parameters.py (hunk 1 above)
Comment (riccardotosi) [reply kratos/tests/test_kratos_parameters.py]: Ok this was my fault, probably doing tests with the serialization I deleted it

File: applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py (hunk 4 above)
Comment (RiccardoTosi) [reply applications/MultilevelMonteCarloApplication/python_scripts/cmlmc_utilities.py]: Shouldn't be better "convergence"? Or it is a Kratos standard to use"convergenced"?

**Comment by riccardotosi** [comment]
I close this PR because I open a new one where I avoid the conflicts I created deleting accidentally the test_model.py and test_kratos_parameters.py. Sorry for the inconvenience.

**Comment by philbucher** [comment]
@RiccardoTosi do you still need this branch?
If not then please delete it

**Pull request closed without merging**